File: frontend/src/lib/currency.ts
import { useEffect, useState } from "react";
import { getExchangeRate, getTaxRate } from "../features/settings/api";

export interface ExchangeRateInfo {
	rate: number;
	// LBP cash denominations in real circulation are large - nobody can
	// make exact change to the nearest lira - so every LBP amount rounds
	// to the nearest multiple of this (admin-configurable in Settings,
	// default 1,000) rather than the raw nearest-lira math.
	rounding: number;
}

// Every price everywhere is stored/entered in USD. This is the one hook
// any feature (Inventory, POS, Sales History, ...) uses to get the
// current admin-set rate (and its rounding rule) and convert on the fly -
// never a second copy of the rate or the math living inside a feature.
export function useExchangeRate(): ExchangeRateInfo | null {
	const [info, setInfo] = useState<ExchangeRateInfo | null>(null);

	useEffect(() => {
		getExchangeRate().then((r) => setInfo({ rate: r.usd_to_lbp_rate, rounding: r.lbp_rounding }));
	}, []);

	return info;
}

// Percentage (7 means 7%) - zero for any install that hasn't set one in
// Settings, so nothing charges tax nobody configured.
export function useTaxRate() {
	const [rate, setRate] = useState<number>(0);

	useEffect(() => {
		getTaxRate().then((r) => setRate(r.tax_rate));
	}, []);

	return rate;
}

export function usdToLbp(amountUsd: number, rate: number, rounding = 1): number {
	const raw = amountUsd * rate;
	return Math.round(raw / rounding) * rounding;
}

export function formatLbp(amountUsd: number, rate: number, rounding = 1): string {
	return `${usdToLbp(amountUsd, rate, rounding).toLocaleString()} LBP`;
}

File: backend/app/shared/currency.py
"""Every price in the system is stored in USD - this is the one place
that converts to LBP, so every domain (Inventory today; POS, Invoicing,
Purchasing later) that needs to show or accept an LBP amount uses the
same rate and the same math, not its own copy of it."""

from sqlalchemy.orm import Session

from ..settings.models import AppSettings, ExchangeRateHistory

DEFAULT_RATE = 89500  # confirmed real-world starting point


def get_or_create_settings(db: Session) -> AppSettings:
	settings = db.get(AppSettings, 1)
	if not settings:
		settings = AppSettings(id=1, usd_to_lbp_rate=DEFAULT_RATE)
		db.add(settings)
		# The very first rate this install ever had is itself a real history
		# entry - without this, a lookup for "what was the rate before the
		# first-ever change" would come back empty.
		db.add(ExchangeRateHistory(usd_to_lbp_rate=DEFAULT_RATE))
		db.commit()
		db.refresh(settings)
	return settings


def usd_to_lbp(amount_usd: float, rate: float, rounding: float = 1) -> float:
	"""rounding defaults to 1 (nearest whole lira) for any caller that
	doesn't pass Settings.lbp_rounding explicitly. LBP cash denominations
	in real circulation are large (1,000s+) - nobody can make change to
	the nearest lira - so anything actually shown to a customer (a
	receipt, an invoice) should pass the admin-configured rounding here,
	same as the frontend's own usdToLbp() (lib/currency.ts) already does."""
	raw = amount_usd * rate
	return round(raw / rounding) * rounding


def lbp_to_usd(amount_lbp: float, rate: float) -> float:
	return round(amount_lbp / rate, 2)
